### PR TITLE
created institute model, spec, factory and updated schema

### PR DESCRIPTION
**Pull request by Luke-kb**
## Description
This PR adds the institute model and basic crud functions with index view and pagination. Institute model and controller specs are added as well as some error_message partial styling for the edit views.

## Breakdown of Commits
Commit 1 : [0ef526a](https://github.com/Luke-kb/mySMS/commit/0ef526a2dc9d67de3cc7d9dc844bb8762bd4d442)
- create institute model
- add institute spec
- migrate and update schema
 
Commit 2 : [65492c1](https://github.com/Luke-kb/mySMS/commit/65492c1f83d1856af2a65e89eb707790f8f6dfc9)
- add institute to dev:populate
- update institute spec

Commit 3 : [c63ec77](https://github.com/Luke-kb/mySMS/commit/c63ec7731a7cd289871278efba6687a74e53c32a)
- added error message partial for edit views

Commit 4 : [d129811](https://github.com/Luke-kb/mySMS/commit/d129811a4bfd57061427750300891d70bb3374d2)
- added institutes_controller with validation
- added routes and index/show/edit views

Commit 5 : [2a95c87](https://github.com/Luke-kb/mySMS/commit/2a95c87f86b11cb8eee5c3c7fd3567e5dc4e6362)
- added institutes_controller specs
- minor tweaks to institutes model and courses controller specs

Commit 6 : [11e03a1](https://github.com/Luke-kb/mySMS/pull/9/commits/11e03a1b869117fd862b7c8517860e9d88dae95d)
- fixed indentation
- refactored institute model spec

Commit 7 : [c81f590](https://github.com/Luke-kb/mySMS/commit/c81f5906489fc5943715524134ed6933b0b55511)
- fixed incorrect comment

#### TODO Checklist
* [x] Create institute model, update schema
* [x] Create institute controller, routes, index/show/edit views and methods
* [x] Create specs for institute
* [x] update dev:populate and institute factory file
* [x] create shared error_message partial for edit forms

### CODE REVIEW

File: app/controllers/institutes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class InstitutesController < ApplicationController
+    before_filter :load_institute, except: [:index]
```
Comment (jbat) [inline app/controllers/institutes_controller.rb]: @Luke-kb looks like some of your settings have changed to 4 spaces ?
The standard style for Ruby is 2 spaces
https://github.com/rubocop/ruby-style-guide#indentation

Comment (jbat) [reply app/controllers/institutes_controller.rb]: As well as a bunch of the files below

File: app/views/shared/_error_messages.html.haml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+- if obj.errors.any?
+    .error_field
+        .error
+            The form contains #{pluralize(obj.errors.count, "error")}.
+        %ul 
+            - obj.errors.full_messages.each do |msg|
+                %li
+                    #{msg}
```
Comment (jbat) [inline app/views/shared/_error_messages.html.haml:8]: Nice one 👍

File: spec/controllers/courses_controller_spec.rb [hunk 3]
```diff
@@ -59,8 +59,6 @@
                 subject
                 expect(course.reload.name).to_not be_nil
             end
-
-
         end
     end
 end
```
Comment (jbat) [inline spec/controllers/courses_controller_spec.rb:64]: Indentation is off in this as well ..

File: spec/models/institute_spec.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      expect(institute).to be_valid
+    end
+  end
+
+  context "with invalid attributes" do
+      let!(:institute) { build(:institute, name: nil) }
+
+      it "is not valid" do
+        institute.valid?
+        expect(institute.errors[:name]).to include("can't be blank")
```
Comment (jbat) [inline spec/models/institute_spec.rb]: do really need to check the error message here.
The presence errors on a models attribute are a core part of ActiveRecord and are already tested as part of ActiveRecords specs.  
In this case .. just a test for is not valid would be enough .. see below

File: spec/models/institute_spec.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      end
+  end
+
```
Comment (jbat) [inline spec/models/institute_spec.rb:16]: dont need this empty line.

Both these contexts in this spec could be cleaned up.
By adding an extra let value for name .. you only have the one build line.
Also the build is only building the object .. not saving it .. so you do not need the bang !
```
  subject { build(:institute, name: name) }

  context 'with valid attributes' do
    let(:name) { "institute1" }

    it { is_expected.to validate_presence_of(:name) }
    it { is_expected.to be_valid }
  end

  context 'with invalid attributes' do
    let(:name) { nil }

    it { is_expected.to_not be_valid }
  end
```

File: app/controllers/institutes_controller.rb (hunk 1 above)
Comment (Luke-kb) [reply app/controllers/institutes_controller.rb]: 😮  thanks for catching that

File: lib/tasks/dev/migrate_title_data.rake [hunk 6]
```diff
@@ -1,19 +1,14 @@
...
-            Student.where(old_title: title.name).update_all(title_id: title.id)
-        end
+  desc 'migrate Student.title data to Title.name column and assign reference id'
+  task migrate_title_data: :environment do
+    # populate Title.name data with unique old_title values 
+    Student.uniq.pluck(:old_title).each do |title|
+        t = Title.find_or_create_by_name(title) if title.present?
+        t.save!
     end
+    # assign title_id to each student in batches
```
Comment (jbat) [inline lib/tasks/dev/migrate_title_data.rake]: @Luke-kb  This comment is no longer applicable.
You are doing an update_all .. not an update of batches with find_by_batch

**Review by jbat — APPROVED** [review]

**Pull request merged**